docs/architecture/clients/dependency-injection.md

```diff
@@ -0,0 +1,219 @@
+---
+sidebar_position: 4
+---
+
+# Dependency Injection
+
+Dependency injection (DI) in our Typescript clients is manually configured across several locations.
+
+You will need to update the DI configuration when you create a new service or change the
+dependencies injected into an existing service.
+
+## Angular and Non-Angular contexts
+
+Our Typescript clients use both Angular and non-Angular contexts. This can vary:
+
+- between apps - for example, the web client is entirely Angular, whereas the CLI client does not
+  use Angular at all
+- within apps - for example, the browser client uses Angular in its popup but not in its background
+  page/service worker
+
+This has important consequences for how we write services and how we configure DI. In particular:
+
+- Angular contexts use the Angular dependency injection framework
+- non-Angular contexts do not use any dependency injection framework and must manually instantiate
+  their dependencies
+
+## Services
+
+Services should be organized as follows depending on where they're used:
+
+- `libs/common` for services used by **multiple clients** in **non-Angular contexts (or a mix of
+  Angular and non-Angular contexts)**
+- `libs/angular` for services used by **multiple clients** in **Angular contexts only**
+- `apps/<client-name>` for services used by a single client only
+
+If a service is used in an Angular context only, it can use the Angular `@Injectable` decorator to
+automatically configure its dependencies. If a service is used in mixed contexts (e.g.
+`libs/common`), it must not use Angular decorators and its dependencies must be manually configured.
+
+## Interfaces
+
+A service should implement an interface defined by an abstract class if:
+
+1. The service will be used across multiple client apps, or
+2. You want to define multiple interfaces for your service to limit what functionality is available
+   to different consumers
+
+For example, you may want to use interfaces to prevent outside modules from calling your internal
+methods:
+
+```ts
+// Injected in other modules
+abstract class MyService {
+  externalMethod: () => void;
+}
+
+// Injected in the MyService feature module
+abstract class InternalMyService extends MyService {
+  internalMethod: () => void;
+}
+
+class MyService implements InternalMyService {
+  externalMethod: () => void;
+  internalMethod: () => void;
+}
+```
+
+If neither of these apply to you, you probably don't need to define an abstract interface for your
+service.
+
+## Configuring DI
+
+### Angular modules
+
+Angular contexts generally use ngModules
+[dependency providers](https://angular.io/guide/dependency-injection-providers) to configure DI. For
+example:
+
```

Comment:
:art: suggestion(non-blocking): Add admonitions and/or consider changing the order of the examples so people don't stop reading after the first code example and think it is the recommended one

```suggestion
Angular contexts generally use ngModules
[dependency providers](https://angular.io/guide/dependency-injection-providers) to configure DI. For
example:

:::danger[BAD]

Do not do this, it is not type safe.

:::

Reply:
This is a good call - it made me wonder whether I shouldn't lead with an example that is wrong, but I think it's necessary to build context. The admonitions help clarify which is the good & bad example though.

Reply:
Yeah I agree that it makes sense to have the bad example first, otherwise you won't know what to look for in the good example. So it's mostly just clarifying that this is one of those "Bad / Good" sections. Here are some  examples I looked at when I wrote this comment:

- Basic text: https://gist.github.com/anichitiandreea/e1d466022d772ea22db56399a7af576b
- Emojis: https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html
- Red code block background: https://google.github.io/styleguide/tsguide.html